src/staking/transactions.ts

```diff
@@ -0,0 +1,752 @@
+import { Psbt, Transaction, networks, payments, script } from "bitcoinjs-lib";
+import { Taptree } from "bitcoinjs-lib/src/types";
+
+import { BTC_DUST_SAT } from "../constants/dustSat";
+import { internalPubkey } from "../constants/internalPubkey";
+import { UTXO } from "../types/UTXO";
+import { PsbtTransactionResult } from "../types/transaction";
+import { isValidBitcoinAddress } from "../utils/btc";
+import { getStakingTxInputUTXOsAndFees, getWithdrawTxFee } from "../utils/fee";
+import { inputValueSum } from "../utils/fee/utils";
+import { buildStakingOutput } from "../utils/staking";
+import { NON_RBF_SEQUENCE, TRANSACTION_VERSION, RBF_SEQUENCE } from "../constants/psbt";
+import { NO_COORD_PK_BYTE_LENGTH } from "../constants/keys";
+
+// https://bips.xyz/370
+const BTC_LOCKTIME_HEIGHT_TIME_CUTOFF = 500000000;
+
+/**
+ * Constructs an unsigned BTC Staking transaction in psbt format.
+ *
+ * Outputs:
+ * - psbt:
+ *   - The first output corresponds to the staking script with the specified amount.
+ *   - The second output corresponds to the change from spending the amount and the transaction fee.
+ *   - If a data embed script is provided, it will be added as the second output, and the fee will be the third output.
+ * - fee: The total fee amount for the transaction.
+ *
+ * Inputs:
+ * - scripts:
+ *   - timelockScript, unbondingScript, slashingScript: Scripts for different transaction types.
+ *   - dataEmbedScript: Optional data embed script.
+ * - amount: Amount to stake.
+ * - changeAddress: Address to send the change to.
+ * - inputUTXOs: All available UTXOs from the wallet.
+ * - network: Bitcoin network.
+ * - feeRate: Fee rate in satoshis per byte.
+ * - publicKeyNoCoord: Public key if the wallet is in taproot mode.
+ * - lockHeight: Optional block height locktime to set for the transaction (i.e., not mined until the block height).
+ *
+ * @param {Object} scripts - Scripts used to construct the taproot output.
+ * such as timelockScript, unbondingScript, slashingScript, and dataEmbedScript.
+ * @param {number} amount - The amount to stake.
+ * @param {string} changeAddress - The address to send the change to.
+ * @param {UTXO[]} inputUTXOs - All available UTXOs from the wallet.
+ * @param {networks.Network} network - The Bitcoin network.
+ * @param {number} feeRate - The fee rate in satoshis per byte.
+ * @param {Buffer} [publicKeyNoCoord] - The public key if the wallet is in taproot mode.
+ * @param {number} [lockHeight] - The optional block height locktime.
+ * @returns {PsbtTransactionResult} The partially signed transaction and the fee.
+ * @throws Will throw an error if the amount or fee rate is less than or equal
+ * to 0, if the change address is invalid, or if the public key is invalid.
+ */
+export function stakingTransaction(
+  scripts: {
+    timelockScript: Buffer;
+    unbondingScript: Buffer;
+    slashingScript: Buffer;
+    dataEmbedScript?: Buffer;
```

Comment:
if I understand correctly this is now the funciton to build phase-2 staking transaction ? If yes, then this ` dataEmbedScript` is not necessary here, as in phase-2 we do not care about anything other than staking output.